File: backend/src/models/reconciliation.py
from typing import Any
from src.models.base import DomainModel, Money


class Reconciliation(DomainModel):
    """Reconciliation domain entity linking a debit and a credit transaction"""

    id: int
    debit_txn_id: int
    credit_txn_id: int
    debit_account_id: str
    credit_account_id: str
    amount: Money
    date_diff_days: int
    match_confidence: float
    match_type: str
    status: str

    @classmethod
    def from_db_row(cls, row: dict[str, Any]) -> "Reconciliation":
        # The `amount` column is stored as a float; normalize to paise (₹1.00 = 100).
        return cls(
            id=row["id"],
            debit_txn_id=row["debit_txn_id"],
            credit_txn_id=row["credit_txn_id"],
            debit_account_id=row["debit_account_id"],
            credit_account_id=row["credit_account_id"],
            amount=Money(paise=row["amount_paise"]),
            date_diff_days=row["date_diff_days"],
            match_confidence=row["match_confidence"],
            match_type=row["match_type"],
            status=row["status"],
        )

File: frontend/lib/mappers/reconciliation.ts
/**
 * Reconciliation Mapper - Pure transformation from DTO to Model
 *
 * This mapper transforms raw API responses to domain-friendly models.
 * It is deterministic, side-effect free, and fully typed.
 */

import type { ReconciliationDto, ReconciliationMatchDto } from '../contracts/api/reconciliation'
import type { ReconciliationModel, ReconciliationMatchModel } from '../models/reconciliation'

/**
 * Map ReconciliationMatch DTO to Model
 *
 * Transformation rules:
 * - Rename fields to camelCase for consistency
 */
export function mapReconciliationMatchToModel(dto: ReconciliationMatchDto): ReconciliationMatchModel {
  return {
    id: dto.id,
    debitTxnId: dto.debit_txn_id,
    creditTxnId: dto.credit_txn_id,
    debitAccountId: dto.debit_account_id,
    creditAccountId: dto.credit_account_id,
    amountPaise: dto.amount_paise,
    dateDiffDays: dto.date_diff_days,
    matchConfidence: dto.match_confidence,
    matchType: dto.match_type,
    status: dto.status,
    createdAt: dto.created_at,
    confirmedAt: dto.confirmed_at,
    // Transaction details
    debitDate: dto.debit_date,
    debitDateIso: dto.debit_date_iso,
    debitDescription: dto.debit_description,
    debitAmountPaise: dto.debit_amount_paise,
    debitBank: dto.debit_bank,
    creditDate: dto.credit_date,
    creditDateIso: dto.credit_date_iso,
    creditDescription: dto.credit_description,
    creditAmountPaise: dto.credit_amount_paise,
    creditBank: dto.credit_bank,
  }
}

/**
 * Map Reconciliation DTO to Model
 *
 * Transformation rules:
 * - Rename fields to camelCase for consistency
 * - Preserve explanation unchanged
 */
export function mapReconciliationToModel(dto: ReconciliationDto): ReconciliationModel {
  return {
    matches: dto.matches.map(mapReconciliationMatchToModel),
    count: dto.count,
    isPartial: dto.is_partial,
    partialReason: dto.partial_reason,
    lastUpdated: dto.last_updated,
    explanation: dto.explanation ?? null,
  }
}

// Re-export types for convenience
export type { ReconciliationModel, ReconciliationMatchModel }